### PR TITLE
Make `always_run` test decorator a tag and improve shard tests.

### LINKED ISSUE #8657: Check test shards do not accidentally overlap
From: https://github.com/numba/numba/pull/8550#discussion_r1043636415_
     
To guard against future mistakes, there should be a check that shards do not overlap anything that is not marked as [always_test](https://github.com/numba/numba/blob/9d4f82df7f0426d89b189eafc7acdc6bf719c716/numba/testing/__init__.py#L64)

### PR DESCRIPTION
**Pull request by stuartarchibald**
This patch:

* Makes an `always_run` test decorator to use on tests that should run in all test shards, it's based on the existing test `tag` mechanism.
* Updates the test runner to recognise the `always_run` tag when sharding so as to ensure that tests tagged with this tag will always be included in a shard.
* Updates the unit tests to ensure fundamental properties with regards to sharding. This includes shard uniqueness, the presence of the `always_run` tests in shards and also that the sum of the test shards and `always_run` tests is the same as the full test listing.
* Fixes up a few references slice->shard.

Fixes #8657

### CODE REVIEW

File: numba/tests/test_runtests.py [hunk 1]
```diff
@@ -119,34 +125,70 @@ def get_count(arg_list):
...
-        sliced = set()
+
+        # Check the tests that should always run are present
+        tmp = self.get_testsuite_listing(['--tag', 'always_test'])
+        always_running = set(self._get_numba_tests_from_listing(tmp))
+        self.assertGreaterEqual(len(always_running), 1)
+
+        for test in always_running:
+            self.assertIn(test, l1)
+            self.assertIn(test, l2)
```
Comment (stuartarchibald) [inline numba/tests/test_runtests.py:148]: This is covered by the test below, is it needed?

File: numba/tests/test_runtests.py [hunk 2]
```diff
@@ -119,34 +125,66 @@ def get_count(arg_list):
...
+        # get some shards
+        sharded = list()
         for i in range(3):
             subset = self.get_testsuite_listing(['-j', '{}:3'.format(i)])
-            sliced.update(filter_test(subset))
-        # The tests must be equivalent
-        self.assertEqual(full, sliced)
+            slist = [*self._get_numba_tests_from_listing(subset)]
+            # there's at least 100 tests in a shard of roughly 1/3!
+            self.assertGreater(len(slist), 100)
```
Comment (apmasell) [inline numba/tests/test_runtests.py:150]: This seems fragile to put in this value.

File: numba/tests/test_runtests.py [hunk 3]
```diff
@@ -119,34 +125,66 @@ def get_count(arg_list):
...
+
+        # check that there is no overlap between the shards
+        for a, b in itertools.combinations(sharded_sets, 2):
+            self.assertFalse(a & b)
+
+        # check that the sum of the shards and the always running tests is the
+        # same as the full listing
+
+        sum_of_parts = set()
+        [sum_of_parts.update(x) for x in sharded_sets]
```
Comment (apmasell) [inline numba/tests/test_runtests.py:181]: Why is this a list comprehension?

File: numba/tests/test_runtests.py (hunk 2 above)
Comment (stuartarchibald) [reply numba/tests/test_runtests.py:150]: The intent here is to make sure that there's not an "accident" whereby the sharding logic has produced an unreasonable quantity of tests in a shard given the size of the test suite. Having seen a few mistakes made through the history of the test listing logic, they tend to manifest as there being be zero or just a few tests listed. As a result I chose 100 as a minimum based on this and as there are 10k+ tests in total it's unlikely a ~1/3 split would ever be less than 100. Also note there are other baked in constants (see `check_all` above which makes sure there are more than 5k tests listed).

Do you think there might be something less "fragile" to check or some other way to assert similar? I suppose querying the size of the test suite and taking a percentage of 1/3 of that might be another way?

**Comment by stuartarchibald** [comment]
@apmasell thanks for the review. Note that since review:
* 3772028 merges `main` and fixes conflicts.
* 228112a fixes a comment `slice`->`shard`.
* 47be4be changes list comp to loops as per review comment.

File: numba/tests/test_runtests.py (hunk 2 above)
Comment (apmasell) [reply numba/tests/test_runtests.py:150]: Take the total and divide by 3. Then it doesn't matter if the test count changes. Or check that the slices are approximately the same size. If they sum to the total and are about equal, then the actual number doesn't matter.

Comment (stuartarchibald) [reply numba/tests/test_runtests.py:150]: Maybe it's best to just remove this assertion. In this module:
* `test_check_shard` asserts that shards are "approximately the same size".
* `test_default`/`test_all` both check that the test suite is being listed as a sensible size (>5000 tests).
* this test checks that the shard sizes add up to the total which is the same as obtained in `test_default`.

I think this renders the assertion under discussion well intentioned but superfluous.

Comment (stuartarchibald) [reply numba/tests/test_runtests.py]: The above is done in b4081c1.

**Review by apmasell — APPROVED** [review]

**Review by sklam — APPROVED** [review]

**Pull request merged**